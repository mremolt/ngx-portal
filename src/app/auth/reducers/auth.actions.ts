import { HttpErrorResponse } from '@angular/common/http';
import { Action } from '@ngrx/store';

export enum AuthActionTypes {
  Authenticate = '[Auth] Authenticate',
  AuthenticateSuccess = '[Auth] Authenticate Success',
  AuthenticateError = '[Auth] Authenticate Error',
  Logout = '[Auth] Logout',
}

export type AuthCredentials = Readonly<{ email: string; password: string }>;

export type AuthToken = Readonly<{
  accessToken: string;
}>;

export class Authenticate implements Action {
  readonly type = AuthActionTypes.Authenticate;
  payload: AuthCredentials;

  constructor(email: string, password: string) {
    this.payload = { email, password };
  }
}

export class AuthenticateSuccess implements Action {
  readonly type = AuthActionTypes.AuthenticateSuccess;

  constructor(public payload: AuthToken) {}
}

export class AuthenticateError implements Action {
  readonly type = AuthActionTypes.AuthenticateError;

  constructor(public payload: HttpErrorResponse) {}
}

export class Logout implements Action {
  readonly type = AuthActionTypes.Logout;
}

export type AuthActions = Authenticate | AuthenticateSuccess | AuthenticateError | Logout;
