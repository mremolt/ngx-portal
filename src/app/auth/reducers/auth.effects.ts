import { HttpClient } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { APP_ENVIRONMENT, ApiError, ApiRequestActionTypes, AppReset, Go } from '@mr/ngx-utils';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { EMPTY, of } from 'rxjs';
import { catchError, flatMap, map, mapTo } from 'rxjs/operators';

import { Environment } from '../../../environments/environment';
import {
  AuthActionTypes,
  AuthToken,
  Authenticate,
  AuthenticateError,
  AuthenticateSuccess,
  Logout,
} from './auth.actions';

@Injectable()
export class AuthEffects {
  @Effect()
  public authenticate$ = this.actions$.pipe(
    ofType<Authenticate>(AuthActionTypes.Authenticate),
    flatMap(action =>
      this.http.post<AuthToken>(`${this.env.apiUrl}/auth/login`, action.payload).pipe(
        map(data => new AuthenticateSuccess(data)),
        catchError(e => of(new AuthenticateError(e)))
      )
    )
  );

  @Effect()
  public redirectAfterLogin$ = this.actions$.pipe(
    ofType<AuthenticateSuccess>(AuthActionTypes.AuthenticateSuccess),
    mapTo(new Go({ path: ['dashboard'] }))
  );

  @Effect()
  public redirectAfterLogout$ = this.actions$.pipe(
    ofType<Logout>(AuthActionTypes.Logout),
    mapTo(new Go({ path: ['/'] }))
  );

  @Effect()
  public resetOnApiErrors$ = this.actions$.pipe(
    ofType<ApiError>(ApiRequestActionTypes.ApiError),
    map(action => {
      if ([401].includes(action.payload.status)) {
        return new AppReset();
      }
      return EMPTY;
    })
  );

  constructor(
    private actions$: Actions,
    private http: HttpClient,
    @Inject(APP_ENVIRONMENT) private env: Environment
  ) {}
}
