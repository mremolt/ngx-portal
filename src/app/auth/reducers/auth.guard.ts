import { Injectable } from '@angular/core';
import { CanActivate, CanActivateChild, CanLoad } from '@angular/router';
import { Store, select } from '@ngrx/store';

import { AuthState, selectAuthenticated } from './auth.selectors';

@Injectable()
export class AuthGuard implements CanActivateChild, CanActivate, CanLoad {
  constructor(private store: Store<AuthState>) {}

  public canActivate() {
    return this.testAuth();
  }

  public canActivateChild() {
    return this.testAuth();
  }

  public canLoad() {
    return this.testAuth();
  }

  private testAuth() {
    return this.store.pipe(select(selectAuthenticated));
  }
}
