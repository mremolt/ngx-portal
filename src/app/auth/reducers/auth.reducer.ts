import { HttpErrorResponse } from '@angular/common/http';

import { AuthActionTypes, AuthActions } from './auth.actions';

export const initialState = {
  loading: false,
  token: { accessToken: '' },
  error: null as HttpErrorResponse | null,
};

export type State = typeof initialState;

export function reducer(state = initialState, action: AuthActions): State {
  switch (action.type) {
    case AuthActionTypes.Authenticate:
      return { ...state, loading: true };

    case AuthActionTypes.AuthenticateSuccess:
      return { ...initialState, token: action.payload };

    case AuthActionTypes.AuthenticateError:
      return { ...initialState, error: action.payload };

    case AuthActionTypes.Logout:
      return initialState;
  }
  return state;
}
