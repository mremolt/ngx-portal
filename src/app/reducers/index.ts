import { RouterAction, RouterReducerState, routerReducer } from '@ngrx/router-store';
import { Action, ActionReducer, ActionReducerMap, MetaReducer } from '@ngrx/store';
import { storeLogger } from 'ngrx-store-logger';

import { environment } from '../../environments/environment';

export interface RootState {
  router: RouterReducerState;
}

export interface AnyAction extends Action {
  payload?: any;
  meta?: object;
}

export function logger(reducer: ActionReducer<RootState>): any {
  return storeLogger({ collapsed: true })(reducer);
}

export const reducers: ActionReducerMap<RootState, RouterAction<any>> = {
  router: routerReducer,
};

export const metaReducers: MetaReducer<RootState>[] = !environment.production ? [logger] : [];
