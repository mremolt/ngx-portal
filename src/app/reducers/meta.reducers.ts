import { resetMetaReducer } from '@mr/ngx-utils';
import { ActionReducer, MetaReducer } from '@ngrx/store';
import { localStorageSync } from 'ngrx-store-localstorage';
import { storeLogger } from 'ngrx-store-logger';

import { environment } from '../../environments/environment';
import { RootState } from './root.reducer';

export function logger(reducer: ActionReducer<RootState>): ActionReducer<RootState> {
  return storeLogger({ collapsed: true })(reducer);
}

export function localStorageSyncReducer(
  reducer: ActionReducer<RootState>
): ActionReducer<RootState> {
  return localStorageSync({ keys: ['auth'], rehydrate: true })(reducer);
}

const commonMetaReducers = [localStorageSyncReducer, resetMetaReducer];

export const metaReducers: MetaReducer<RootState>[] = environment.production
  ? commonMetaReducers
  : [logger, ...commonMetaReducers];
