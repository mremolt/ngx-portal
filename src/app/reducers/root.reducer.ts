import { RouterAction, RouterReducerState, routerReducer } from '@ngrx/router-store';
import { ActionReducerMap } from '@ngrx/store';

export interface RootState {
  router: RouterReducerState;
}

export const reducers: ActionReducerMap<RootState, RouterAction<any>> = {
  router: routerReducer,
};
