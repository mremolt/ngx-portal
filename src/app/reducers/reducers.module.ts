import { CommonModule } from '@angular/common';
import { InjectionToken, NgModule } from '@angular/core';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, META_REDUCERS, MetaReducer, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { environment } from 'src/environments/environment.prod';

import { metaReducers } from './meta.reducers';
import { RootState, reducers } from './root.reducer';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<RootState>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

export function getMetaReducers(): MetaReducer<RootState>[] {
  return metaReducers;
}

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forRoot(REDUCER_TOKEN),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    StoreRouterConnectingModule.forRoot(),
  ],
  providers: [
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: META_REDUCERS,
      useFactory: getMetaReducers,
    },
  ],
})
export class ReducersModule {}
