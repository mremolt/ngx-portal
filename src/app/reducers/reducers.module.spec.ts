import { ReducersModule, getMetaReducers, getReducers } from './reducers.module';

describe('ReducersModule', () => {
  let reducersModule: ReducersModule;

  beforeEach(() => {
    reducersModule = new ReducersModule();
  });

  it('should create an instance', () => {
    expect(reducersModule).toBeTruthy();
  });

  describe('getReducers', () => {
    it('returns the reducers', () => {
      expect(getReducers()).toMatchSnapshot();
    });
  });

  describe('getMetaReducers', () => {
    it('returns an Array of meta reducers', () => {
      const subject = getMetaReducers();
      expect(subject).toEqual(expect.any(Array));

      expect(subject).toMatchSnapshot();
    });
  });
});
