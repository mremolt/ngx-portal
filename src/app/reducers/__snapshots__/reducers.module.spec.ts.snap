// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReducersModule getMetaReducers returns an Array of meta reducers 1`] = `
Array [
  [Function],
  [Function],
  [Function],
]
`;

exports[`ReducersModule getReducers returns the reducers 1`] = `
Object {
  "router": [Function],
}
`;
